crypto/evp/m_sigver.c

```diff
@@ -208,7 +208,13 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
                                           mdname, provkey, params);
     }
 
-    goto end;
+    /*
+     * If the operation was not a success and no digest was found, an error
+     * needs to be raised.
+     */
+    if (ret > 0 || mdname != NULL)
+        goto end;
+    ERR_raise(ERR_LIB_EVP, EVP_R_NO_DEFAULT_DIGEST);
```

Comment:
That reason should only be given if `type == NULL`

Reply:
It can't be.  The check @line 150 precludes this.  Unless someone registers a digest with the name _UNDEF_.

I can add a check for `type == NULL` here even though it's redundant.

Reply:
I'll have you look at lines 155-161, and ask you if you believe it's impossible for that to produce a NULL `mdname`.

Reply:
That is **exactly** the place I'm trying to catch.  It isn't possible to raise the error there because some algorithms legitimately do not have a default digest and there is no way to tell if this is the case or not.  Raising the error here after the init call has failed at least indicates the probable cause of the failure.

Reply:
[sigh]

*The point is* that if those lines produce a NULL `mdname`, *then you should have* `type == NULL && mdname == NULL`
No?  What am I missing?

Reply:
What happens when the digest_verify_init or digest_sign_init call fails?  It is definitely possible for `mdname != NULL && type ==NULL` so checking just `type` by itself is wrong.  Checking both is redundant.

`mdname == NULL` only when the code has gone through the lines 155-161 and hasn't found a default digest.  That is the case this error needs to be raise on.

Reply:
Huh?  I never said you should check `type` alone.  You already do check `mdname` two lines up, did I ask you to remove that?  No I didn't.

However, as far as I can see, you may have two cases in this line:

- `mdname == NULL && type == NULL` (that would be when lines 155-161 gave a NULL `mdname`)
- `mdname == NULL && type != NULL` (that would be when `canon_mdname()` returned NULL on line 153

Reply:
Look at `canon_mdname()`, it won't return NULL when passed the name of the digest in `type`:

https://github.com/openssl/openssl/blob/ef1e0242a9aec5210845df86162c0b9219ff0f11/crypto/evp/m_sigver.c#L27-L37

Reply:
Does this make things clearer?

``` C
    /*
     * If the operation was not a success and no digest was found, an error
     * needs to be raised.
     */
    if (ret > 0 || mdname != NULL)
        goto end;
    if (type == NULL)
        ERR_raise(ERR_LIB_EVP, EVP_R_NO_DEFAULT_DIGEST);
```

Reply:
It does.  I still argue that the `type == NULL` check is redundant.

Nerf guns at dawn at 15Mm.

Reply:
Hmm, ok, I get what you say...  'cause the supposition is that `EVP_MD_get0_name()` will never return NULL.

I would rather still have that check, even if it's redundant *with current code*.  Just the fact that we've had this dispute shows that it's too easy to forget this if we modify code higher up, and mess things up.

Reply:
Added

Reply:
Thank you